altair/utils/save.py

```diff
@@ -9,28 +9,26 @@
 
 
 def write_file_or_filename(
-    fp: Union[str, pathlib.PurePath, IO],
+    fp: Union[pathlib.Path, IO],
```

Comment:
What's the reason to use `Path` instead of `PurePath`? In my understanding, `PurePath` is more generic as `Path` is a subclass of it.

Reply:
I spent some time thinking about this, because it was my first time seeing `PurePath` usage outside of the python docs. So I did a re-read to try and understand why `PurePath` may have originally been used here.

Near the top of the [pathlib docs](https://docs.python.org/3/library/pathlib.html) it states:
> Pure paths are useful in some special cases; for example:
> 1. If you want to manipulate Windows paths on a Unix machine (or vice versa). You cannot instantiate a [WindowsPath](https://docs.python.org/3/library/pathlib.html#pathlib.WindowsPath) when running on Unix, but you can instantiate [PureWindowsPath](https://docs.python.org/3/library/pathlib.html#pathlib.PureWindowsPath).
> 2. You want to make sure that your code only manipulates paths without actually accessing the OS. In this case, instantiating one of the pure classes may be useful since those simply don’t have any OS-accessing operations.

I don't believe either of those cases are relevant here.

From my understanding the core difference between `PurePath` and `Path` is that the latter introduces methods for I/O. Either type can be passed to `open`, as they are [PathLike](https://docs.python.org/3/library/os.html#os.PathLike) - but I believe using `Path` more clearly communicates that it will be used for I/O. 

> In my understanding, PurePath is more generic as Path is a subclass of it. @binste 

If the choice is based on wanting more generic option, `Path` already handles this internally based on the host OS: 
![pathlib-inheritance](https://github.com/vega/altair/assets/125183946/4f238dc6-7e64-44a4-bfe4-6ad3db8b9a02)

I will note, there are changes [scheduled for python 3.13](https://discuss.python.org/t/make-pathlib-extensible/3428/168) that will implement a more traditional ABC hierarchy, that is documented [here](https://pathlib-abc.readthedocs.io/en/latest/).

Reply:
@binste as the remaining comments all tie into `Path`/`PurePath`, I'll wait to hear back from you on this and then make the remaining changes in one commit. 

Really appreciate you taking the time to review

Reply:
And I appreciate that you take the time to thoroughly research it and provide such an explanation! :)

I see your point about `Path` better communicating that it will be used for I/O. I've also never seen PurePath before, when adding type hints to this method I think i took it over from some other part of the library but I can't see a good reason for it anymore and maybe it's easier to just stick to what people know.

-> I'm ok with using `Path` instead of `PurePath` as you have already implemented. Would still be great to keep `str` around as an input argument for the functions where it was possible to use before.